staging/src/k8s.io/client-go/transport/round_trippers.go

```diff
@@ -113,6 +113,45 @@ func (rt *authProxyRoundTripper) RoundTrip(req *http.Request) (*http.Response, e
 	return rt.rt.RoundTrip(req)
 }
 
+// legalHeaderRune reports whether r is a valid rune in a header
+// field name. RFC 7230 says:
+//   header-field   = field-name ":" OWS field-value OWS
+//   field-name     = token
+//   tchar = "!" / "#" / "$" / "%" / "&" / "'" / "*" / "+" / "-" / "." /
+//           "^" / "_" / "`" / "|" / "~" / DIGIT / ALPHA
+//   token = 1*tchar
+func legalHeaderRune(r rune) bool {
+	return int(r) < len(legalHeaderRunes) && legalHeaderRunes[r]
+}
+
+func percentEncodeByte(b byte) string {
+	return fmt.Sprintf("%%%x", b)
+}
+
+func percentEncodeRune(r rune) string {
+	bytes := []byte(string(r))
+	encoded := ""
+	for _, b := range bytes {
+		encoded += percentEncodeByte(b)
+	}
+	return encoded
+}
+
+func percentEncodeIfIllegal(r rune) string {
+	if legalHeaderRune(r) {
+		return string(r)
+	}
+	return percentEncodeRune(r)
+}
+
+func headerEscape(key string) string {
+	encoded := ""
+	for _, r := range key {
+		encoded += percentEncodeIfIllegal(r)
```

Comment:
Does this require a lot of allocations? Should we use a byte buffer?

Reply:
potentially...
I deliberately chose string concatenation because bytes.Buffer.WriteFoo() returns an error, and I thought that it would be weird to do error handling when such a case does not arise from string concatenation (which I assume `panic`s when malloc or something fails)

Reply:
Scratch that, the docs are explicit in that Write(Byte|Rune|String) always returns a nil error. I'll use the buffer.

Reply:
Use strings.Builder

Reply:
done